cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -517,36 +524,52 @@ public void upgradeApplications(FullHttpRequest request, HttpResponder responder
         for (ApplicationId appId : appIds) {
-          ApplicationUpdateDetail updateDetail;
-          try {
-            applicationLifecycleService.upgradeApplication(appId, allowedArtifactScopes, allowSnapshot);
-            updateDetail = new ApplicationUpdateDetail(appId);
-          } catch (UnsupportedOperationException e) {
-            String errorMessage = String.format("Application %s does not support upgrade.", appId);
-            updateDetail = new ApplicationUpdateDetail(appId, new NotImplementedException(errorMessage));
-          } catch (InvalidArtifactException | NotFoundException e) {
-            updateDetail = new ApplicationUpdateDetail(appId, e);
-          } catch (Exception e) {
-            updateDetail =
+          Future<?> future = executor.submit(() -> {
+            ApplicationUpdateDetail updateDetail;
+            try {
+              applicationLifecycleService.upgradeApplication(appId, allowedArtifactScopes, allowSnapshot);
+              updateDetail = new ApplicationUpdateDetail(appId);
+            } catch (UnsupportedOperationException e) {
+              String errorMessage = String.format("Application %s does not support upgrade.", appId);
+              updateDetail = new ApplicationUpdateDetail(appId, new NotImplementedException(errorMessage));
+            } catch (InvalidArtifactException | NotFoundException e) {
+              updateDetail = new ApplicationUpdateDetail(appId, e);
+            } catch (Exception e) {
+              updateDetail =
                 new ApplicationUpdateDetail(appId, new ServiceException("Upgrade failed due to internal error.", e,
-                                            HttpResponseStatus.INTERNAL_SERVER_ERROR));
-            LOG.error("Application upgrade failed with exception", e);
-          }
-          GSON.toJson(updateDetail, ApplicationUpdateDetail.class, jsonWriter);
-          jsonWriter.flush();
-          chunkResponder.sendChunk(Unpooled.wrappedBuffer(outputStream.toByteArray()));
-          outputStream.reset();
-          chunkResponder.flush();
+                                                                        HttpResponseStatus.INTERNAL_SERVER_ERROR));
+              LOG.error("Application upgrade failed with exception", e);
+            }
+
+            lock.lock();
+            try {
+              GSON.toJson(updateDetail, ApplicationUpdateDetail.class, jsonWriter);
+              jsonWriter.flush();
+              chunkResponder.sendChunk(Unpooled.wrappedBuffer(outputStream.toByteArray()));
+              outputStream.reset();
+              chunkResponder.flush();
+            } catch (Exception e) {
+              throw new RuntimeException(e);
+            } finally {
+              lock.unlock();
+            }
+          });
+          futureList.add(future);
+        }
+        executor.shutdown();
+        for (Future<?> future : futureList) {
+          future.get();
```

Comment:
If the task throw exception, I think it is better to catch and log it here and move on to the next future, otherwise the response to the client will be in broken json.

Reply:
I'm not sure if this is an expected behavior. According to [original version](https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java#L545), it will throw exception directly.